From 9df6708be4467ef9efd6e3b5979e74e3b0f0aa53 Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Wed, 04 May 2022 10:31:52 +0300
Subject: [PATCH] [SW-82717] rom: add axprot settings

As per the gaudi3 security specifications axprot of jtag, i2c,
smi and pcie_elbi is required to be set to secured in order for
the range register configuration to take effect, as these engines
are by default privileged.

Change-Id: Ic952506b93eb919c5afa452072bd1088db5f6c8c
---

diff --git a/arch/gaudi3/arc/security.c b/arch/gaudi3/arc/security.c
index 7253a04..ae01925 100644
--- a/arch/gaudi3/arc/security.c
+++ b/arch/gaudi3/arc/security.c
@@ -28,7 +28,14 @@
 #define RR_LBW_MAX_SHORT_RR			14
 #define RR_LBW_PRIV_RANGE_MAX_FULL_VAL		0x1FFFFFFC
 #define RR_HIT_WR_TO_RD_DIST			8
-
+#define PROT_OVR_SECURE				0
+#define PROT_OVR_PRIV				1
+#define PROT_OVR_UNPRIV				0
+#define PROT_OVR_USER				1
+#define AXPROT_OVRD_EN				BIT(0)
+#define AXPROT_OVRD_VAL_POS			4
+#define AXPROT_SEC_VAL		((PROT_OVR_SECURE << AXPROT_OVRD_VAL_POS) | AXPROT_OVRD_EN)
+#define AXPROT_UNPRIV_VAL	((PROT_OVR_UNPRIV << AXPROT_OVRD_VAL_POS) | AXPROT_OVRD_EN)
 /*
  * This key hash is calculated on the Public Key Properties -
  *   -----------------------------------------------
@@ -142,6 +149,18 @@
 	regmap_writel(RR_PRIV_RANGE_EN, mmD0_I2C_S_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
 }
 
+static void security_i2c_axprot_config(void)
+{
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_LBW_AWPROT_0_OVRD);
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_LBW_ARPROT_0_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_LBW_AWPROT_1_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_LBW_ARPROT_1_OVRD);
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_AWPROT_0_OVRD);
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_ARPROT_0_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_AWPROT_1_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_I2C_S_MSTR_IF_AXPROT_HBW_ARPROT_1_OVRD);
+}
+
 static void security_enable_i2c_s_if(void)
 {
 	/*
@@ -152,6 +171,8 @@
 	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_I2C_S_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_13);
 	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_I2C_S_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
 	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_I2C_S_MSTR_IF_RR_HBW_PRIV_RANGE_EN_7);
+
+	security_i2c_axprot_config();
 }
 
 static void security_disable_smi_if(void)
@@ -171,6 +192,18 @@
 	regmap_writel(RR_PRIV_RANGE_EN, mmD0_SMI_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
 }
 
+static void security_smi_axprot_config(void)
+{
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_SMI_MSTR_IF_AXPROT_LBW_AWPROT_0_OVRD);
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_SMI_MSTR_IF_AXPROT_LBW_ARPROT_0_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_SMI_MSTR_IF_AXPROT_LBW_AWPROT_1_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_SMI_MSTR_IF_AXPROT_LBW_ARPROT_1_OVRD);
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_SMI_MSTR_IF_AXPROT_HBW_AWPROT_0_OVRD);
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_SMI_MSTR_IF_AXPROT_HBW_ARPROT_0_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_SMI_MSTR_IF_AXPROT_HBW_AWPROT_1_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_SMI_MSTR_IF_AXPROT_HBW_ARPROT_1_OVRD);
+}
+
 static void security_enable_smi_if(void)
 {
 	/*
@@ -181,6 +214,8 @@
 	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_SMI_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_13);
 	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_SMI_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
 	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_SMI_MSTR_IF_RR_HBW_PRIV_RANGE_EN_7);
+
+	security_smi_axprot_config();
 }
 
 void security_enable_twi_if(void)
@@ -234,6 +269,18 @@
 	regmap_writel(RR_PRIV_RANGE_EN, mmD0_JT_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
 }
 
+static void security_jtag_axprot_config(void)
+{
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_JT_MSTR_IF_AXPROT_LBW_AWPROT_0_OVRD);
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_JT_MSTR_IF_AXPROT_LBW_ARPROT_0_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_JT_MSTR_IF_AXPROT_LBW_AWPROT_1_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_JT_MSTR_IF_AXPROT_LBW_ARPROT_1_OVRD);
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_JT_MSTR_IF_AXPROT_HBW_AWPROT_0_OVRD);
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_JT_MSTR_IF_AXPROT_HBW_ARPROT_0_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_JT_MSTR_IF_AXPROT_HBW_AWPROT_1_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_JT_MSTR_IF_AXPROT_HBW_ARPROT_1_OVRD);
+}
+
 void security_enable_jtag_if(void)
 {
 	security_jtag_dap(SECURITY_JTAG_DAP_ENABLE);
@@ -246,6 +293,8 @@
 	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_JT_MSTR_IF_RR_LBW_PRIV_RANGE_EN_SHORT_13);
 	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_JT_MSTR_IF_RR_HBW_PRIV_RANGE_EN_0);
 	regmap_writel(RR_PRIV_RANGE_DIS, mmD0_JT_MSTR_IF_RR_HBW_PRIV_RANGE_EN_7);
+
+	security_jtag_axprot_config();
 }
 
 void security_handle_jtag_if(void)
@@ -915,8 +964,17 @@
 	}
 }
 
+static void security_pcie_elbi_axprot_config(void)
+{
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_PCIE_ELBI_RR_MSTR_IF_AXPROT_LBW_AWPROT_0_OVRD);
+	regmap_writel(AXPROT_UNPRIV_VAL, mmD0_PCIE_ELBI_RR_MSTR_IF_AXPROT_LBW_ARPROT_0_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_PCIE_ELBI_RR_MSTR_IF_AXPROT_LBW_AWPROT_1_OVRD);
+	regmap_writel(AXPROT_SEC_VAL, mmD0_PCIE_ELBI_RR_MSTR_IF_AXPROT_LBW_ARPROT_1_OVRD);
+}
+
 void security_rr_recovery_config(void)
 {
+	security_pcie_elbi_axprot_config();
 	security_prot_bits_config();
 	security_rr_recovery_lbw_config();
 	security_rr_recovery_hbw_config();
